server/sonar-web/src/main/js/apps/code/utils.js

```diff
@@ -0,0 +1,175 @@
+  }
+}
+
+function prepareChildren (r) {
+  return {
+    components: r.components,
+    total: r.paging.total,
+    page: r.paging.pageIndex
+  };
+}
+
+function skipRootDir (breadcrumbs) {
+  return breadcrumbs.filter(component => {
+    return !(component.qualifier === 'DIR' && component.name === '/');
+  });
+}
+
+function storeChildrenBase (children) {
+  children.forEach(addComponent);
+}
+
+function storeChildrenBreadcrumbs (parentComponentKey, children) {
+  const parentBreadcrumbs = getComponentBreadcrumbs(parentComponentKey);
+  if (parentBreadcrumbs) {
+    children.forEach(child => {
+      const breadcrumbs = [...parentBreadcrumbs, child];
+      addComponentBreadcrumbs(child.key, breadcrumbs);
+    });
+  }
+}
+
+export function retrieveComponentBase (componentKey) {
+  const existing = getComponentFromBucket(componentKey);
+  if (existing) {
+    return Promise.resolve(existing);
+  }
+
+  return getComponent(componentKey, METRICS_WITH_COVERAGE).then(component => {
+    addComponent(component);
+    return component;
+  });
+}
+
+function retrieveComponentChildren (componentKey) {
+  const existing = getComponentChildren(componentKey);
+  if (existing) {
+    return Promise.resolve({
+      components: existing.children,
+      total: existing.total
+    });
+  }
+
+  return getChildren(componentKey, METRICS_WITH_COVERAGE, { ps: PAGE_SIZE, s: 'name' })
+      .then(prepareChildren)
+      .then(expandRootDir)
+      .then(r => {
+        addComponentChildren(componentKey, r.components, r.total);
+        storeChildrenBase(r.components);
+        storeChildrenBreadcrumbs(componentKey, r.components);
+        return r;
+      });
+}
+
+function retrieveComponentBreadcrumbs (componentKey) {
+  const existing = getComponentBreadcrumbs(componentKey);
+  if (existing) {
+    return Promise.resolve(existing);
+  }
+
+  return getBreadcrumbs({ key: componentKey })
+      .then(skipRootDir)
+      .then(breadcrumbs => {
+        addComponentBreadcrumbs(componentKey, breadcrumbs);
+        return breadcrumbs;
+      });
+}
+
+export function retrieveComponent (componentKey) {
+  return Promise.all([
+    retrieveComponentBase(componentKey),
+    retrieveComponentChildren(componentKey),
+    retrieveComponentBreadcrumbs(componentKey)
+  ]).then(r => {
+    return {
+      component: r[0],
+      components: r[1].components,
+      total: r[1].total,
+      page: r[1].page,
+      breadcrumbs: r[2]
+    };
+  });
+}
+
+export function loadMoreChildren (componentKey, page) {
+  return getChildren(componentKey, METRICS_WITH_COVERAGE, { ps: PAGE_SIZE, p: page })
+      .then(prepareChildren)
+      .then(expandRootDir)
+      .then(r => {
+        addComponentChildren(componentKey, r.components, r.total);
+        storeChildrenBase(r.components);
+        storeChildrenBreadcrumbs(componentKey, r.components);
+        return r;
+      });
+}
+
+export function parseError (error) {
+  try {
+    return error.response.json().then(r => {
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Remove curly braces and "return" from this arrow function body. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=javascript%3AS3524)